buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -57,6 +58,74 @@
     return new LinkedList<String>(Arrays.asList(TOKENIZER_PATTERN.split(input)));
   }
 
+  /**
+   * Adds specified keywords and nonReservedKeywords to extractedData. Also
+   * ensures that nonReservedKeywords is a subset of the union of keywords
+   * and extractedData.keywords.
+   *
+   * @param keywords The keywords to add
+   * @param nonReservedKeywords the non reserved keywords to add
+   * @param extractedData The object to which the keywords will be added to
+   *
+   * @throws IllegalStateException when an an element in nonReservedKeywords is
+   *                               not in keywords or extractedData.keywords
+   */
+  public void processKeywords(Map<Keyword, String> keywords,
+      Set<Keyword> nonReservedKeywords,
+      ExtractedData extractedData) throws IllegalStateException {
+    for (Keyword keyword : nonReservedKeywords) {
+      if (!(keywords.containsKey(keyword)
+            || extractedData.keywords.containsKey(keyword))) {
```

Comment:
Why does the existence of "keyword" in extractedData.keywords result in a IllegalStateException(" is not a keyword.") exception being thrown?

Perhaps these should be split out into two separate exception/error messages?

Reply:
I think you may have missed the negation there, the behaviour is that if a nonReservedKeyword is not in keywords and not in extractedData.keywords an exception should be thrown.

Reply:
You're right, I did miss the parenthesized negation :)

nit: In that case, it might be a bit more readable without the extra parentheses. For example:

if (!keywords.containsKey(keyword) && !extractedData.keywords.containsKey(keyword))

Reply:
Sounds good, will make the change.